2021q4/vdso.adoc

```diff
@@ -0,0 +1,79 @@
+Main reason why it is possible is the often mentioned co-evolution of
+the kernel and C runtime.  We can naturally introduce features that
+require implementation both in kernel, and support in the userspace
+parts, since FreeBSD is developed as the whole.  Surprisingly, it also
+allows for kernel and dynamic linker to know much less (and not
+enforce anything) about userspace consumers of interfaces.
+
+For instance, syscall-less wall clock was implemented long time ago,
+by kernel providing time hands blob in the shared page, and C library
+knowing about the blob location and supported algorithms.  There is no
+need for VDSO that interposes some libc symbols, or provides services
+that are named by known symbols to userspace.
+
+From the all years of experience with this pseudo-VDSO approach, the
+only feature that was arguably impossible to implement without
+providing real VDSO, was the signal trampoline dwarf annotations, for
+the benefit of stack unwinders.
+
+When kernel delivers signal to userspace, it changes some key
+registers (like instruction pointer, stack pointer, whatever is needed
+by architecture), and pushes the saved image of whole usermode CPU
+state (context) into user stack.  Then, the control is passed to the
+small piece of code located on the shared page (signal trampoline),
+which calls into user function-handler, and on return from the handler
+issues sigreturn(2) syscall to reload old context.
+
+From this description, it is clear that the state of the machine
+during trampoline execution is quite different from the normal C
+calling frames.  Unwinders that handle things like C++ exceptions,
+Rust panics, or other similar mechanisms in specific language
+runtimes, need to understand the specialness of the trampoline frame.
+The current approach is to hardcode the detection of the trampoline,
+e.g. by matching the instruction pointer against sysctl
+kern.proc.sigtramp.
+
+DWARF annotations are enough to provide the required information to
+unwinders to make trampoline frame not special anymore, but the
+problem is that there is no way for unwinders to find the annotation
+without introducing even more speciality.  Or, we can inject VDSO
+which only purpose is to appear in enumeration of the dso's loaded
+into process, either with dl_iterate_phdr(3) (in-process) or r_debug
+(remote), with PT_GNU_EH_FRAME header pointing to the root of DWARF
+info.
+
+This is exactly what VDSO on FreeBSD does: it wraps signal trampoline
+bits and their DWARF annotation (.cfi) into shared object, which is
+put into shared page, and which is linked by rtld into set of
+preloaded objects on the image activation.
+
+Efforts were made to strip as much as possible of unneeded structures
+and paddings from the VDSO image, because it consumes space in the
+shared page.  It was pushed as far as the common denominator of lld
+and ld.bfd allowed, with several tricks done by linker scripts, and
+some use of seemingly undocumented linker options.
```

Comment:
Please clarify "seemingly undocumented". Do you mean "poorly documented", "undocumented", "obscure", or something else?

Reply:
I was able to find needed options by reading lld source code.

Reply:
If they're not in the documentation, they're undocumented. No "seemingly" about it.

Reply:
I am not an expert in locating lld documentation.  The options I used might be explained somewhere, but I do not know the place.  I prefer a weaker statement there.